google-cloud-clients/google-cloud-firestore/src/test/java/com/google/cloud/firestore/it/ITSystemTest.java

```diff
@@ -889,95 +893,184 @@ public void onEvent(
+    //      test. These changes are expected to trigger a number of events that should be delivered
+    //      to the snapshot listener from 1.
+    //
+    // The mechanics of how the test is executed are as follows
+    //   A. All events delivered to the listener in 1 are added to a list of event we receive.
+    //   B. A separate ExecutorService is created and has tasks submitted to it to perform the
+    //      the series of document updates mentioned in 2. (Each document gets its own task)
+    //   C. CountDownLatches are used to keep track of completion of work while the test is running.
+    //      * Every time an event is received by the listener in 1 a CDL is decremented.
+    //      * Each document update task has a CDL that is decremented when all actions have been
+    //        performed
+    //      * Each CDL has a timeout associated with its await (this is important so that in the
+    //        case there is a hang during the test the whole suite isn't taken with it)
+    //   D. After all updates have been performed and the expected number of events have been
+    //      received assertions on the events will be performed.
     ListenerRegistration registration = null;
+    final ExecutorService updatesExecutor = Executors.newCachedThreadPool();
 
-    final Set<String> expectedEvents =
-        Sets.newTreeSet(
-            newHashSet("event 0", "event 1", "event 2", "event 3", "event 4", "event 5"));
-    final Set<String> actualEvents = Collections.synchronizedSet(new TreeSet<String>());
+    final List<ListenerEvent> receivedEvents =
+        Collections.synchronizedList(new ArrayList<ListenerEvent>());
 
     try {
+      // create our CDLs that are used to track work completion
+      final CountDownLatch doc1CDL = new CountDownLatch(1);
+      final CountDownLatch doc2CDL = new CountDownLatch(1);
+      final CountDownLatch eventsCDL = new CountDownLatch(6);
+
+      final Query query = randomColl.whereEqualTo("foo", "bar");
+      // register the snapshot listener for the query
       registration =
-          randomColl
-              .whereEqualTo("foo", "bar")
-              .addSnapshotListener(
-                  new EventListener<QuerySnapshot>() {
-                    DocumentReference ref1;
-                    DocumentReference ref2;
-
-                    @Override
-                    public void onEvent(
-                        @Nullable QuerySnapshot value, @Nullable FirestoreException error) {
-                      System.out.printf("onEvent(value : %s, error : %s)%n", value, error);
-                      try {
-                        switch (semaphore.availablePermits()) {
-                          case 0:
-                            actualEvents.add("event 0");
-                            assertTrue(value.isEmpty());
-                            ref1 = randomColl.add(map("foo", "foo")).get();
-                            ref2 = randomColl.add(map("foo", "bar")).get();
```

Comment:
I think the problem with this test is that these two adds are not guaranteed to generate two different events. One way of fixing this would be either split these up into two tasks or use a WriteBatch for both writes, which is guaranteed to only raise one event of size 2.

Reply:
This scenario feels a bit like a different test case to me. I can look at adding another test if that sounds good to you?

Reply:
It's different for sure - but this test is just meant to show that basic Watch functionality works. The actual steps that we verify are of minor importance.